config-svc/pkg/api/routes_v1.go

```diff
@@ -15,93 +15,137 @@
+func (s *Server) PostClustersAdd(ctx echo.Context) error {
+	var data v1.PostClustersAddJSONRequestBody
 	if err := ctx.Bind(&data); err != nil {
 		return err
 	}
 
-	labels := data.Labels.AdditionalProperties
-	if labels == nil {
-		labels = make(map[string]string)
+	secure := false
+	if data.CouchbaseConfig.Secure != nil && *data.CouchbaseConfig.Secure {
+		secure = true
 	}
-	if data.NameLabel != nil {
-		labels[*data.NameLabel] = data.Name
-	}
-	value := prometheusTargets{
-		{
-			Targets: data.Targets,
-			Labels:  labels,
-		},
+	mgmtPort := 8091
+	if data.CouchbaseConfig.ManagementPort != nil {
+		mgmtPort = int(*data.CouchbaseConfig.ManagementPort)
 	}
-	if value[0].Labels == nil {
-		value[0].Labels = make(map[string]string)
-	}
-	valueJSON, err := json.Marshal(value)
+	cluster, err := couchbase.FetchCouchbaseClusterInfo(
+		data.Hostname,
+		mgmtPort,
+		secure,
+		data.CouchbaseConfig.Username,
+		data.CouchbaseConfig.Password,
+	)
 	if err != nil {
 		return err
 	}
 
-	var path string
-	if s.production {
-		path = prometheusTargetsPathProduction
-	} else {
-		path = prometheusTargetsPathDevelopment
+	staticConfig := prometheus.StaticConfig{
+		Targets: make([]string, len(cluster.Nodes)),
+		Labels: map[string]string{
+			"cluster": cluster.ClusterName,
+		},
 	}
-	path = fmt.Sprintf(path, sanitize.Name(data.Name))
 
-	if data.Overwrite == nil || !*data.Overwrite {
-		_, err := os.Stat(path)
-		if err == nil {
-			return echo.NewHTTPError(http.StatusConflict, "target already exists")
-		} else if errors.Is(err, fs.ErrNotExist) {
-			// Ignore it
-		} else {
+	var anyNodeCB7 bool
+
+	for i, node := range cluster.Nodes {
+		hostname, mgmtPort, err := node.ResolveHostPort(secure)
+		if err != nil {
 			return err
 		}
+		if node.Version.AtLeast(cbvalue.Version7_0_0) {
+			staticConfig.Targets[i] = fmt.Sprintf("%s:%d", hostname, mgmtPort)
+			anyNodeCB7 = true
+		} else {
+			// TODO: allow customising this
+			staticConfig.Targets[i] = fmt.Sprintf("%s:%d", hostname, 9091)
+		}
 	}
 
-	outFile, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o664)
+	cfgPath := os.Getenv("PROMETHEUS_CONFIG_FILE")
+	if cfgPath == "" {
+		cfgPath = defaultPrometheusConfigPath
+	}
+	cfgFile, err := os.OpenFile(cfgPath, os.O_RDWR, 0)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to open Prometheus config: %w", err)
 	}
-	if _, err = outFile.Seek(0, 0); err != nil {
-		return err
+	defer cfgFile.Close()
+	existingConfig, err := io.ReadAll(cfgFile)
+	if err != nil {
+		return fmt.Errorf("failed to read Prometheus config: %w", err)
 	}
-	if _, err = outFile.Write(valueJSON); err != nil {
-		return err
+	var cfg prometheus.Configuration
+	if err := yaml.Unmarshal(existingConfig, &cfg); err != nil {
+		return fmt.Errorf("failed to parse Prometheus config: %w", err)
+	}
+
+	scrapeConfig := prometheus.ScrapeConfig{
+		// Job name needs to be unique
+		JobName:       fmt.Sprintf("couchbase-server-managed-%d", len(cfg.ScrapeConfigs)+1),
```

Comment:
Presumably thread safe here?

Reply:
We operate on our local copy of the file, and we `os.OpenFile` it with `os.O_RDWR`, so (if I've understood correctly) the OS will lock the file for us

Reply:
Testy McTestface then would be good if we need to. I'm not sure the control path around this call - is it always correct is my main question when you're relying on the length of the array to set something else, can the array size be changed elsewhere whilst this is going on? I don't think so but wanted to confirm.

Reply:
It can't, `cfg` is a local variable to this function that isn't passed to anything else.

I'm not sure how we'd test the locking - and really at that point aren't we testing "Linux does sensible things when opening files"?